platform/cog-platform-drm.c

```diff
@@ -236,6 +279,180 @@ init_egl ()
 }
 
 
+static void
+input_handle_key_event (struct libinput_event_keyboard *key_event)
+{
+    struct wpe_input_xkb_context *default_context = wpe_input_xkb_context_get_default ();
+    struct xkb_state *state = wpe_input_xkb_context_get_state (default_context);
+
+    uint32_t key = libinput_event_keyboard_get_key (key_event) + 8;
+    uint32_t keysym = xkb_state_key_get_one_sym (state, key);
+    uint32_t unicode = xkb_state_key_get_utf32 (state, key);
+
+    struct wpe_input_keyboard_event event;
+    event.time = libinput_event_keyboard_get_time (key_event);
+    event.key_code = keysym;
+    event.hardware_key_code = unicode;
+
+    enum libinput_key_state key_state = libinput_event_keyboard_get_key_state (key_event);
+    event.pressed = (key_state == LIBINPUT_KEY_STATE_PRESSED);
+
+    event.modifiers = 0;
+
+    wpe_view_backend_dispatch_keyboard_event (wpe_view_data.backend, &event);
+}
+
+static void
+input_handle_touch_event (enum libinput_event_type touch_type, struct libinput_event_touch *touch_event)
+{
+    uint32_t time = libinput_event_touch_get_time (touch_event);
+
+    enum wpe_input_touch_event_type event_type = wpe_input_touch_event_type_null;
+    switch (touch_type) {
+    case LIBINPUT_EVENT_TOUCH_DOWN:
+        event_type = wpe_input_touch_event_type_down;
+        break;
+    case LIBINPUT_EVENT_TOUCH_UP:
+        event_type = wpe_input_touch_event_type_up;
+        break;
+    case LIBINPUT_EVENT_TOUCH_MOTION:
+        event_type = wpe_input_touch_event_type_motion;
+        break;
+    case LIBINPUT_EVENT_TOUCH_FRAME:
+    {
+        struct wpe_input_touch_event event;
+        event.touchpoints = input_data.touch_points;
+        event.touchpoints_length = 10;
+        event.type = input_data.last_touch_type;
+        event.id = input_data.last_touch_id;
+        event.time = time;
+        event.modifiers = 0;
+
+        wpe_view_backend_dispatch_touch_event (wpe_view_data.backend, &event);
+
+        for (int i = 0; i < 10; ++i) {
```

Comment:
This hardcoded array length here is bugging me. We should add this somewhere:

```c
#define N_ELEMENTS(arr)  ((sizeof (arr) / sizeof (arr[0]))
```

and then use `i < N_ELEMENTS (input_data.touch_points)` as the loop condition.

Reply:
Actually, we are already using GLib which has [G_N_ELEMENTS](https://developer.gnome.org/glib/stable/glib-Standard-Macros.html#G-N-ELEMENTS:CAPS)—please use that.